src/schemes/fluid/surface_normal_sph.jl

```diff
@@ -121,30 +133,140 @@ function remove_invalid_normals!(system::FluidSystem, surface_tension::SurfaceTe
     return system
 end
 
-function remove_invalid_normals!(system, surface_tension)
-    # Normal not needed
+# see Morris 2000 "Simulating surface tension with smoothed particle hydrodynamics"
+function remove_invalid_normals!(system::FluidSystem,
+                                 surface_tension::Union{SurfaceTensionMorris,
+                                                        SurfaceTensionMomentumMorris},
+                                 surfn::ColorfieldSurfaceNormal)
+    (; cache, smoothing_length, smoothing_kernel, ideal_neighbor_count) = system
+    (; ideal_density_threshold, interface_threshold) = surfn
+
+    # We remove invalid normals i.e. they have a small norm (eq. 20)
+    normal_condition2 = (interface_threshold /
+                         compact_support(smoothing_kernel, smoothing_length))^2
+
+    for particle in each_moving_particle(system)
+
+        # heuristic condition if there is no gas phase to find the free surface
+        # We remove normals for particles which have alot of support e.g. they are in the inside
+        if ideal_density_threshold > 0 &&
+           ideal_density_threshold * ideal_neighbor_count < cache.neighbor_count[particle]
+            cache.surface_normal[1:ndims(system), particle] .= 0
+            continue
+        end
+
+        particle_surface_normal = cache.surface_normal[1:ndims(system), particle]
```

Comment:
First allocation here:
```suggestion
        particle_surface_normal = surface_normal(system, particle)
```